Readme.md
# z80dismblr

z80dismblr is a Z80 command line disassembler written in typescript.

I.e. it is complete and working in general but it is still very young so there might be some faults still.
Please don't hesitate to report any problem that you find.

## Features

- Supports binary and *.sna (snapshot) files.
- Can read MAME trace (*.tr) files for better results.
- Supports undocumented opcodes.
- Supports Spectrum Next opcodes.
- Graphical visualization via Graphivz .dot of
  - call graphs
  - flow-charts
- Disassembles the given binary via Code-Flow-Graph analysis.
- Divides into data and code area.
- Separates subroutines visually.
- Creates labels from hex addresses.
- Uses "local" label syntax inside subroutines.
- Points out all callers/callees of a subroutine.
- Customization of the output
  - Label prefixes
  - List file with or without address and opcode bytes
  - Opcodes in upper or lower case


## Installation

<!--
You can either install the npm z80dismblr package or use the provided executables directly.

npm install:
~~~
$ npm install z80dismblr
~~~

If you want to use the executable instead you find executables for Windows, Mac and Linux [here](https://github.com/maziac/z80dismblr/releases). Simply download and execute from the command line.
-->

At the moment there is no npm install package available but you can install the executable directly.
Executables exist for Windows, Mac and Linux.
Just download [here](https://github.com/maziac/z80dismblr/releases), unzip and execute from the command line.


## Usage

Simply execute the unzipped file from the shell.

_Note: Usage is shown here for MacOS only, it should work similar for Linux and Windows._

To create an assembler listing for the snapshot file 'myfile.sna' just use:
~~~
$ ./z80dismblr-macos --sna myfile.sna --out myfile.list
~~~

It reads in the file (which is in SNA file format) and writes it to the 'myfile.list' file.

For example the disassembly looks like this:
~~~
; Data is referenced by 5 locations:
; 9011h(in SUB164), 9048h(in SUB167), 901Ch(in SUB166), 9031h(in SUB166), 9018h(in SUB165)
8FED DATA146:
8FED 00           DEFB 0      ; 00h

...

; Subroutine: Size=38, CC=4.
; Called by: INTRPT1[A612h].
; Calls: SUB164.
901C SUB166:
901C 2A ED 8F     LD   HL,(DATA146) ; 8FEDh
901F .sub166_loop:
901F 7E           LD   A,(HL)
9020 FE FF        CP   255    ; FFh,   -1
9022 28 11        JR   Z,.sub166_l2 ; 9035h
9024 11 00 30     LD   DE,12288 ; 3000h
9027 CB 77        BIT  6,A
9029 28 01        JR   Z,.sub166_l1 ; 902Ch
902B 1C           INC  E
902C .sub166_l1:
902C 12           LD   (DE),A
902D 32 00 10     LD   (DATA003),A ; 1000h
9030 23           INC  HL
9031 22 ED 8F     LD   (DATA146),HL ; 8FEDh
9034 C9           RET
9035 .sub166_l2:
9035 2A EF 8F     LD   HL,(DATA148) ; 8FEFh
9038 23           INC  HL
9039 23           INC  HL
903A CD 0A 90     CALL SUB164 ; 900Ah
903D 7C           LD   A,H
903E B7           OR   A
903F 20 DE        JR   NZ,.sub166_loop ; 901Fh
9041 C9           RET
~~~

A SNA file contains an entry point into the code. So it is not necessary to provide a'--codelabel'.
However, the entry point in the SNA file might not be very good for disassembly purposes in that case
prepare more information via '--codelabel'.


You can also read in binary files (without headers), e.g. MAME roms.
For binary files you have to provide additional info of the offset address
of the loaded file.
~~~
$ ./z80dismblr-macos --bin 0 rom1.bin --bin 0x1000 rom2.bin --bin 0x2000 rom3.bin --codelabel 0x800 MAIN_START --out roms.list
~~~
This will load 3 binary files (rom1.bin, rom2.bin and rom3.bin).
rom1.bin starts at address 0, rom2.bin at address 0x1000 and rom3.bin at address 0x2000.
There are 2 initial labels where code starts: at 0x800 the main program starts. Address 0 is added automatically as program start.

If you know nothing about the code the better way will be to provide a MAME trace file. I.e. you run MAME with the debugger and the trace option
and save it to a file, e.g. myfile.tr.
Now you start a disassembly and provide the file:
~~~
$ ./z80dismblr-macos --bin 0 rom1.bin --bin 0x1000 rom2.bin --bin 0x2000 rom3.bin --tr myfile.tr --out roms.list
~~~
Note that you can but you don't have to provide a --codelable in this case.


You can higly customize the appearance of the output, e.g. you can suppress the shown address or the opcode bytes.

~~~
; Subroutine is referenced by 1 location:
; 0xe1f1(in SUB305)
SUB156:
             bit  0,e
             jr   z,.sub156_l ; 8E67h
             ld   hl,36353 ; 8E01h, -29183
             ld   (hl),c
             inc  hl
             ld   (hl),b
.sub156_l:
             call sub157 ; 8E73h
             ld   h,141  ; 8Dh, -115
             ld   l,a
.sub156_loop:
             ld   (hl),e
             inc  hl
             dec  d
             jr   nz,.sub156_loop ; 8E6Dh
             ret
~~~


Please use
~~~
$ ./z80dismblr-macos -h
~~~
to print a help for all allowed arguments.


### Arguments File

Instead of a big argument list you can also pass all arguments via a file.
The format is exactly the same as on the commandline.
To disassemble the rom file from above you would need a file with
the following contents:
~~~
$ cat argsfile
--bin 0 rom1.bin
--bin 0x1000 rom2.bin
--bin 0x2000 rom3.bin
--tr myfile.tr
~~~

~~~
$ ./z80dismblr-macos --args argsfile --out roms.list
~~~


## Statistics

Apart from the disassembly output with the labels and the mnemonics z80dismblr also prints out a few statistics in the comments.
For each subroutines it lists the callers and callees.
Additional the size of the subroutine is shown in bytes and the cyclomatic complexity (CC).

~~~
; Subroutine: Size=27, CC=1.
; Called by: SUB300[D74Ch], SNA_LBL_MAIN_START_A660[CBA0h].
; Calls: SUB042, SUB046, SUB077, SUB086, SUB244, SUB245.
D72E SUB299:
~~~

## Visualization

### Caller Graphs

With the '--callgraphout' option it is possible to let z80dismblr create .dot files for use with [Graphviz](http://www.graphviz.org).

Here is an example for the program "Star Warrior" (48K ZX Spectrum). Use z80dismblr like this:
~~~
$ ./z80dismblr-macos --sna starwarrior.sna --out starwarrior.list --callgraphout starwarrior.dot
~~~

It will generate the 'starwarrior.dot' file from the SNA file.
If you look at the dot file with Graphviz it will look like this:

![](documentation/images/starwarrior_dot.jpg)

Although this looks very confusing on first sight a few things can be learned from this view:

- We get an overview of all subroutines and how there are interconnected. Each arrow means: subroutine "SUBn" calls subroutine "SUBm".
- Each bubble represents a subroutine (or entry point). It contains the name, its size in bytes and its cyclomatic complexity.
- The size of the bubble is related to its size in bytes. I.e. bigger subroutines lead to bigger bubbles.
- We can see the leafs, i.e. the subroutines that do not call other subroutines. Often these are very generic functions like math calculations etc. When doing reverse engineering it is often helpful to start with those functions and work from bottom to top to understand the
higher layer subroutines.
- We can see one or more roots, e.g. the main routine. We can also try a top-down analysis to understand the called subroutines.
- Calls into unassigned memory (i.e. addresses outside of the given binary) are shown in gray.


The highlighted roots:

![](documentation/images/starwarrior_dot_root.jpg)
![](documentation/images/starwarrior_dot_root2.jpg)
![](documentation/images/starwarrior_wrong_sub.jpg)

This example shows 4 roots. Why is this?
1. SNA_LBL_MAIN_START_A660 is the address from the SNA file. Since no other code parts reference (jumps to) it, it is a root. Here truly the program starts.
2. INTRPT1 is the interrupt that is called 50 times per second on the Spectrum.
Normally z80dismblr cannot find interrupts because it uses a CFG analysis and if no location refers to the interrupt z80dismblr cannot see it. So you would have to manually set the interrupt address via an argument to z80dimblr ("--codelabel address"). In this case however the "-tr" option was used and so z80dismblr could additionally analyse the traces and find the interrupt by itself.
3. INTRPT2: This in fact is the real interrupt location. Here a simple "JP INTERPT1" could be found. The reason why z80dismblr did not draw any lines from here is: it is self-modifying code. The binary that z80dismblr analysed simply contains 3 "NOP" operations. Thus there is no label. The jump operation and the jump location is written by executing the code. But since z80dismblr doesn't do a dynamic analysis it cannot see the these values.
4. SUB007: This looks strange. And indeed, this helped me to find an error in the assembler program. It was hard to find but in the end the code boiled down to the very simple:
~~~
711D:
        ...
        CALL NZ,711D
        ...
~~~
I.e. a recursive call to itself which was wrong coding simply. I didn't intend to write a recursive function.
However z80dismblr thinks 711D is a subroutine because it is called via a CALL so it assigns the 711D as a label ("SUB007"). But no other location refers to 711D so that the address has no caller, i.e. no arrow pointing to it.
z80dismblr will spit out a warning now in cases like the one above:
~~~
$ Warning: Address: 711Dh. A subroutine was found that calls itself recursively but is not called from any other location.
~~~
In the dot graphic the subroutine is highlighted by a different color.

---


A leaf (i.e. a subroutine that does not call any other subroutine):

![](documentation/images/starwarrior_dot_leaf.jpg)


A call to unassigned memory result in a gray bubble (in case of SNA files for the ZX Spectrum this would be calls into the ROM area):

![](documentation/images/starwarrior_dot_equ.jpg)


#### Sub Call Graphs

It is also possible to let z80dismblr generate only a part of the caller graphs e.g. to focus on a certain subroutine.

For this add "--noautomaticaddr" to the commandline. This will prevent that z80dimblr will use address 0000 or the SNA start address automatically.

Additionally add the address of the subroutine you want to see with a "--codelabel" option:
~~~
$ ./z80dismblr-macos --sna starwarrior.sna --callgraphout starwarrior.dot --noautomaticaddr --codelabel 0x735E SUB19
~~~
You can additionally add a label name (here we chose "SUB19" so that it is the same name as in the big caller graph diagram).
You can get the address from the previously created 'starwarrior.list' file.

The result is a call graph just for the subroutine at address 0x753E:

![](documentation/images/starwarrior_sub19.jpg)



### Flow Charts

Via the '--flowchart...' arguments it is possible to create flowcharts of subroutines.

With '--flowchartout filename' you specify the output path. The generated file is a .dot file that can be visualized with [Graphviz](http://www.graphviz.org).

with '--flowchartaddresses addr1 addr2 ... addrN' you can specify one or more subroutines that you want to visualize.

E.g.:
~~~
$ ./z80dismblr-macos --sna starwarrior.sna --flowchartout fc.dot --flowchartaddresses 7015h A3ABh
~~~
will create the following graph:

![starwarrior_fc_dot](documentation/images/starwarrior_fc_dot.jpg)

## "Interactive" Usage

During reverse engineering of a binary at first very little is known about the code.
Then after looking at the disassembly the one or the other subroutine is understood and can be commented with more senseful comments than the one that z80dismblr generates.

Therefore you can input a file with labels and comments via the '--lblsin file' option.

The file is read and substitutes the label name and the comments for a given address.

Here is a small real world example of a printing subroutine.
The original disassembled code:
~~~
; Subroutine: Recursive, Size=44, CC=4.
; Called by: SUB443[D752h], SUB442[D734h], SUB178[8405h], SUB179[8439h], SUB005[706Dh], SUB359[CB27h], self[7631h], SUB006[70A9h], SUB004[7054h].
; Calls: SUB039, SUB055, SUB103.
760A SUB055:
760A 7E           ld   a,(hl)
760B 23           inc  hl
760C FE FF        cp   FFh    	; 255,   -1
760E C8           ret  z
760F FE FE        cp   FEh    	; 254,   -2
7611 28 09        jr   z,.sub055_l1 	; 761Ch
7613 FE FD        cp   FDh    	; 253,   -3
7615 28 15        jr   z,.sub055_l2 	; 762Ch
7617 CD 44 74     call SUB039 	; 7444h
761A 18 EE        jr   SUB055 	; 760Ah
761C .sub055_l1:
761C 7E           ld   a,(hl)
761D 23           inc  hl
761E E5           push hl
761F 66           ld   h,(hl)
7620 6F           ld   l,a
7621 7E           ld   a,(hl)
7622 23           inc  hl
7623 66           ld   h,(hl)
7624 6F           ld   l,a
7625 CD 24 7B     call SUB103 	; 7B24h
7628 .sub055_loop:
7628 E1           pop  hl
7629 23           inc  hl
762A 18 DE        jr   SUB055 	; 760Ah
762C .sub055_l2:
762C 7E           ld   a,(hl)
762D 23           inc  hl
762E E5           push hl
762F 66           ld   h,(hl)
7630 6F           ld   l,a
7631 CD 0A 76     call SUB055 	; 760Ah
7634 18 F2        jr   .sub055_loop 	; 7628h
~~~

After analysing we find out what the purpose is and how it works so we add comments in a special file:
~~~
; Subroutine to print a text in HL until an end-of-string (0xFF) is found.
; There is a little formatting allowed:
; FEh: the next 2 bytes are interpreted as a number
; FDh: the next 2 bytes are interpreted as a pointer to another text
; (which could also contain formatting.)
760a sub_print_formatted_text_hl

760c ; End of string

760f ; integer (%d)

7613 ; string (%s)
~~~

This results in the more readable disassembly:
~~~
; Subroutine to print a text in HL until an end-of-string (0xFF) is found.
; There is a little formatting allowed:
; FEh: the next 2 bytes are interpreted as a number
; FDh: the next 2 bytes are interpreted as a pointer to another text
; (which could also contain formatting.)
760A sub_print_formatted_text_hl:
760A 7E           ld   a,(hl)
760B 23           inc  hl
760C FE FF        cp   FFh    	; End of string
760E C8           ret  z
760F FE FE        cp   FEh    	; integer (%d)
7611 28 09        jr   z,.sub_print_formatted_text_hl_l1 	; 761Ch
7613 FE FD        cp   FDh    	; string (%s)
7615 28 15        jr   z,.sub_print_formatted_text_hl_l2 	; 762Ch
7617 CD 44 74     call SUB039 	; 7444h
761A 18 EE        jr   sub_print_formatted_text_hl 	; 760Ah
761C .sub_print_formatted_text_hl_l1:
761C 7E           ld   a,(hl)
761D 23           inc  hl
761E E5           push hl
761F 66           ld   h,(hl)
7620 6F           ld   l,a
7621 7E           ld   a,(hl)
7622 23           inc  hl
7623 66           ld   h,(hl)
7624 6F           ld   l,a
7625 CD 24 7B     call SUB102 	; 7B24h
7628 .sub_print_formatted_text_hl_loop:
7628 E1           pop  hl
7629 23           inc  hl
762A 18 DE        jr   sub_print_formatted_text_hl 	; 760Ah
762C .sub_print_formatted_text_hl_l2:
762C 7E           ld   a,(hl)
762D 23           inc  hl
762E E5           push hl
762F 66           ld   h,(hl)
7630 6F           ld   l,a
7631 CD 0A 76     call sub_print_formatted_text_hl 	; 760Ah
7634 18 F2        jr   .sub_print_formatted_text_hl_loop 	; 7628h
~~~

Please note:
- The label for the address is used now not only at address 760A but also in all locations that jump to or call this address. See e.g. address 761A or 7631
- The addresses 7D0C, 760F and 7613 get an inline comment which explains roughly the purpose of the comparison.



## Recommendations

If you know nothing about the binary that you disassemble the output of the z80dismblr might be disappointing.

CHANGELOG.md
## 1.6.0
- Opcodes refactored and fixed.
- Testcases added for all opcodes.


## 1.5.2
- REG_VERSION string conversion fixed.
- Changed >> to >>>.
- Merged with DeZog.


## 1.5.1
- Fixed disassembly of ZX Next opcode 'PUSH nn'.
- Updated libraries.


## 1.5.0
- Renamed '--lblsin' to '--comments'.
- Fix: Labels in dot file now support dot-notation.
- New parameter '--rstend address' to stop the disassembler diving into the RST subroutine. Use e.g. '--rstend 8' to get correct disassembly results if ESXDOS file handling is used.
- Added parameter '--callgraphnode addr|label' to output only a certain label in the dot graph.
- Added parameter '--callgraphformat format' to offer more format options for dot.
- Renamed old parameter '--callgraphformat format' to '--callgraphnodeformat format'.
- '--callgraphhighlight' now also allows labels as input.


## 1.4.2
- Corrected opcode "SUB A,s" to "SUB s".
- Added new Z80N barrel shift and "JP (C)" opcodes.
- Corrected "JP (IX)" and "JP (IY)" disassembly.
- Made "JP (HL)" etc. a stopping insruction.


## 1.4.1
- Improved decoding of 'NEXTREG': Register names are decoded.
- Internal changes for z80-debug.


## 1.4.0
- "--opcode": User opcode extensions
- Branch labels that jump into the middle of an opcode adjusted. These use offsets now (e.g. "JP LBL1+1").
- Warnings added for wrong branch labels.
- Warnings added for code that is accessed as data (self-modifying code).
- Corrected handling of apostrophes in argument files.


## 1.3.0
- Output of flowcharts:
  "--flowchartout" and "--flowchartaddresses: Output flow-chart for a particular address (subroutine).
- "--dot..." renamed to "--callgraph...".


## 1.2.0
- "--dotformat": Formatting of dot node output.
- Changed "--noaddr0" to "--noautomaticaddr". Also suppresses the SNA start address.
- Subroutines are divided into several subroutines if not in a coherent block.
- Opcodes use hex numbers now, comments decimal.
- "--lblsin": Input of labels with comments.


## 1.1.0
- Supports output of callgraphs into dot files for visualization with graphviz.
- Callgraph visualization emphasizes:
  - Code start
  - Leafs
  - Interrupt
- Improved trace file parsing.
- Better interrupt recognition.


## 1.0.0
- Disassembles the given binary via Code-Flow-Grap analysis.
- Divides into data and code area.
- Creates labels from hex addresses.
- Distinguishes labels for subroutines or jump addresses.
- Separates subroutines visually.
- Uses "local" label syntax inside subroutines.
- Points out all callers of a subroutine.
- Customization of the output
- Label prefixes
- List file with or without address and opcode bytes
- Opcodes in upper or lower case
- Can read MAME trace (*.tr) files for better results.
- Supports *.sna (snapshot) files.
- Supports undocumented opcodes.
- Supports Spectrum Next opcodes.


## 0.2.0
- Support for mame .tr (trace) files.

## 0.1.0
Initial version.


src/main.ts
import { Disassembler } from './disasm';
import { readFileSync } from 'fs';


class Startup {

    /// String containing all whitespaces.
    protected static whiteSpaces = ' \t\n\r';

    /// The disassembler instance.
    protected static dasm = new Disassembler();

    /**
     * Main function. Called on startup.
     */
    public static main(): number {
        try {
            // Set defaults
            this.dasm.opcodesLowerCase = true;
            this.dasm.addOpcodeBytes = false;

            // Get arguments
            const args = process.argv.splice(2);

            // Check for help
            if(args.length == 0) {
                this.printHelp();
                return 0;
            }

            // Go through arguments
            this.processArgs(args);

            // Execute
            const lines = this.dasm.disassemble();

            // Temporary output
            console.log(lines.slice(0,20000).join('\n'));


        }
        catch(e) {
            console.log(e);
            return 1;
        }

        return 0;
    }


    /**
     * Prints command line help.
     */
    protected static printHelp() {
        console.log(`
z80dismblr is a disassembler for Z80 code binaries. The disassembly will be
simply written to stdout.

Example usages:

$ z80dismblr --sna myfile.sna > myfile.list
This will write the disassembly of the snapshot file 'myfile.sna' to file 'myfile.list'.

$ z80dismblr --bin 0x8000 myfile.obj --codelable 0x9000 > myfile.list
This will write the disassembly of the binary file file 'myfile.obj' to file 'myfile.list'.
The binary file starts at address 0x8000 and the code entry point start at
0x9000 (i.e. here begins the code area).

General usage:
z80dismblr [options]
    -h|-help|--help Print this help.
    --args file: Instead of an argument list on the command line it is also possible
        to provide the arguments in a file. There is no special format, just
        provide the arguments. May conatin newlines.
    --sna file: Read in a snapshot file. (Snapshot files contain a code start
        address.)
    --bin start file: Read in a plain binary. 'start' is the address in memory for the read binary.
        --codelabel or --tr is mandatory to
        obtain any disassembly results.
    --tr file: Add a MAME trace file. This can be used instead of --codelabel.
        Providing a trace file will increase the quality of the
        disassembly output.
    --codelabel address [labelname]: Known addresses (like the code start
        address) can be given here. If no sna file is given at least one
        codelabel is required so  that disassembly can start from that address.
    --jmptable address size: If it is known that a jump-table exists in memory
        then its address and size can be given here. 'size' is the number of addresses.

    Prefixes: It is possible to customize the label naming. There are different
    types of labels. For each label type you can define its prefix.
        --subprefix prefix: Prefix for subroutines, e.g. '--subprefix MYSUB'.
        --lblprefix prefix: Prefix for adresses reached by a 'JP'.
        --rstprefix prefix: Prefix for adresses reached via a 'RST'.
        --datalblprefix prefix: Prefix for data areas.
        --locallblprefix prefix: Prefix for local lable, those reached with a (positive)
        'JR'. Note that in front of this prefix the subroutines main prefix is added.
        --localloopprefix prefix: Similar to the local lables but for negative relative jumps.
        --selfmodprefix prefix: For selfmodifying address, default is 'SELF_MOD'.

    Formatting options:
        --clmnsaddress value: The size of the address field at the beginning of the line. If 0 no address is shown. Otherwise the address is shown in hex, so senseful numbers start at 4.
        --clmnsbytes value: The length used for the opcode bytes.
        --clmnsopcodefirstpart value: The size of the first part of the opcode, e.g. 'LD'.
        --clmsnOpcodeTotal value: The size of the complete opcode, e.g. 'LD  A,(HL)'.
        --uppercase: Use upper case for the opcodes, e.g. 'ld a,(hl)'.
        --addbytes: Print also the byte values of the opcodes (the opcode bytes).
    `);
    }


    /**
     * Processes the command line arguments or the arguments read from a file.
     * @param args List of arguments.
     */
    protected static processArgs(args: Array<string>) {
        // Iterate all arguments
        let arg;
        let path;
        let addressString;
        let addr;
        let text;
        let value;
        while(arg = args.shift()) {
            // Check option
            switch(arg) {
                // Help
                case '--help':
                case '-help':
                case '-h':
                    this.printHelp();
                    return 0;

                // arguments file
                case '--args':
                    path = args.shift();
                    if(!path) {
                        throw arg + ': No path given.';
                    }
                    // Read arguments from file
                    const fileArgs = this.readArgsFromFile(path);
                    // Process
                    this.processArgs(fileArgs);
                    break;

                // SNA file
                case '--sna':
                    path = args.shift();
                    if(!path) {
                        throw arg + ': No path given.';
                    }
                    this.dasm.readSnaFile(path);
                    break;

                // SNA file
                case '--bin':
                    // get origin
                    const originString = args.shift();
                    const origin = this.parseValue(originString);
                    if(isNaN(origin)) {
                        throw arg + ": Not a number: " + originString;
                    }
                    // get path
                    path = args.shift();
                    if(!path) {
                        throw arg + ': No path given.';
                    }
                    // set memory
                    this.dasm.memory.readBinFile(origin, path);
                    break;

                // TRACE (.tr) file
                case '--tr':
                    path = args.shift();
                    if(!path) {
                        throw arg + ': No path given.';
                    }
                    this.dasm.useMameTraceFile(path);
                    break;

                // set a code label
                case '--codelabel':
                    // parse address
                    addressString = args.shift();
                    addr = this.parseValue(addressString);
                    if(isNaN(addr)) {
                        throw arg + ": Not a number: " + addressString;
                    }
                    // Now optional name
                    let labelName = args.shift();
                    if(labelName)
                        if(labelName.startsWith('--')) {
                            // is the next arguments
                            args.unshift(labelName);
                            labelName = undefined;
                        }
                    // Set label
                    this.dasm.setLabel(addr, labelName);
                    break;

                // set a jump table
                case '--jmptable':
                    // parse address
                    addressString = args.shift();
                    addr = this.parseValue(addressString);
                    if(isNaN(addr)) {
                        throw arg + ": Not a number: " + addressString;
                    }
                    // Now the size
                    const sizeString = args.shift();
                    const size = this.parseValue(sizeString);
                    if(isNaN(size)) {
                        throw arg + ": Not a number: " + sizeString;
                    }
                    // Set jump table
                    this.dasm.setJmpTable(addr, size);
                    break;


                // Prefixes

                // SUB prefix
                case '--subprefix':
                    // get prefix
                    text = args.shift();
                    if(!text || text.length == 0)
                        throw arg + ": No prefix given";
                    // Set prefix
                    this.dasm.labelSubPrefix = text;
                    break;

                // LBL prefix
                case '--lblprefix':
                    // get prefix
                    text = args.shift();
                    if(!text || text.length == 0)
                        throw arg + ": No prefix given";
                    // Set prefix
                    this.dasm.labelLblPrefix = text;
                    break;

                // RST prefix
                case '--rstprefix':
                    // get prefix
                    text = args.shift();
                    if(!text || text.length == 0)
                        throw arg + ": No prefix given";
                    // Set prefix
                    this.dasm.labelRstPrefix = text;
                    break;

                // Data label prefix
                case '--datalblprefix':
                    // get prefix
                    text = args.shift();
                    if(!text || text.length == 0)
                        throw arg + ": No prefix given";
                    // Set prefix
                    this.dasm.labelDataLblPrefix = text;
                    break;

                // Local label prefix
                case '--locallblprefix':
                    // get prefix
                    text = args.shift();
                    if(!text || text.length == 0)
                        throw arg + ": No prefix given";
                    // Set prefix
                    this.dasm.labelLocalLablePrefix = text;
                    break;

                // Local loop prefix
                case '--localloopprefix':
                    // get prefix
                    text = args.shift();
                    if(!text || text.length == 0)
                        throw arg + ": No prefix given";
                    // Set prefix
                    this.dasm.labelLoopPrefix = text;
                    break;

                // Self modifying prefix
                case '--selfmodprefix':
                    // get prefix
                    text = args.shift();
                    if(!text || text.length == 0)
                        throw arg + ": No prefix given";
                    // Set prefix
                    this.dasm.labelSelfModifyingPrefix = text;
                    break;


                // Formatting

                // Start lines with address
                case '--clmnsaddress':
                    // get value
                    text = args.shift();
                    value = this.parseValue(text);
                    if(isNaN(value))
                       throw arg + ": Wrong value " + text;
                    this.dasm.clmnsAddress = value;
                    break;

                // The opcode bytes
                case '--clmnsbytes':
                    // get value
                    text = args.shift();
                    value = this.parseValue(text);
                    if(isNaN(value))
                       throw arg + ": Wrong value " + text;
                    this.dasm.clmnsBytes = value;
                    break;

                // The size of the first part of the opcode, e.g. 'LD'
                case '--clmnsopcodefirstpart':
                    // get value
                    text = args.shift();
                    value = this.parseValue(text);
                    if(isNaN(value))
                       throw arg + ": Wrong value " + text;
                    this.dasm.clmnsOpcodeFirstPart = value;
                    break;

                // The size of the complete opcode, e.g. 'LD  A,(HL)'
                case '--clmsnOpcodeTotal':
                    // get value
                    text = args.shift();
                    value = this.parseValue(text);
                    if(isNaN(value))
                       throw arg + ": Wrong value " + text;
                    this.dasm.clmsnOpcodeTotal = value;
                    break;

                // Use upper case for opcodes
                case '--uppercase':
                    this.dasm.opcodesLowerCase = false;
                    break;

                // print also ocode byte values
                case '--addbytes':
                    this.dasm.addOpcodeBytes = true;
                    break;

                default:
                    throw "Unknown argument: '" + arg + "'";
                    return 1;
            }
        }
    }

    /**
     * Reads argument list from a file.
     * @param path The path to the file.
     * @return List of strings, like args.
     */
    protected static readArgsFromFile(path: string): Array<string> {
        const argsData = readFileSync(path).toString();
        const args = new Array<string>();
        let k = 0;
        while(true) {
            // skip whitespaces
            k = this.skipWhiteSpaces(argsData, k);

            // Check if first character is a ' or "
            const val = argsData[k];
            let l;
            if(val == '"' || val == "'") {
                // Search for ending
                k ++;
                l = argsData.indexOf(val, k);
                if(l < 0) {
                    throw "Couldn't find closing " + val + " in " + path;
                }
                // get string
                const arg = argsData.substr(k, l-k);
                args.push(arg);
                // next
                l ++;
            }
            else {
                // search for next whitespace (== ending)
                l = this.findWhiteSpace(argsData, k+1);
             }

             // get string
             const arg = argsData.substr(k, l-k);
             args.push(arg);

             // next
             k = l;
        }

        // return
        return args;
    }


    /**
     * Skips white spaces.
     * @param s The string to search.
     * @param i The start index.
     * @return The index of the first non-whitespace.
     */
    protected static skipWhiteSpaces(s: string, i: number): number {
        const len = s.length;
        while(i < len) {
            const val = s[i];
            if(this.whiteSpaces.indexOf(val) < 0)
                break;
            // next
            i ++;
        }
        return i;
    }


    /**
     * Finds the next white spaces.
     * @param s The string to search.
     * @param i The start index.
     * @return The index of the first whitespace.
     */
    protected static findWhiteSpace(s: string, i: number): number {
        const len = s.length;
        while(i < len) {
            const val = s[i];
            if(this.whiteSpaces.indexOf(val) >= 0)
                break;
            // next
            i ++;
        }
        return i;
    }


    /**
	 * Parses a string and converts it to a number.
	 * The string might be decimal or in an hex format.
	 * If the string begins with '0x' or '$' or ends with 'h' or 'H'
	 * it is assumed to be a hex value.
	 * If the string ends with 'b' or 'B' a bit value is assumed.
	 * Otherwise decimal is used.
	 * If the string starts with _ a flag value is assumed. I.e. following flags
	 * are allowed: SZHPNC
	 * Otherwise decimal is used.
	 * @param valueString The string to convert. Ignores case.
	 * @returns The value of valueString. Can also return NaN in error cases.
	 */
	public static parseValue(valueString: string|undefined): number {
        if(!valueString)
            return NaN;

		const match = /^\s*((0x|\$)([0-9a-f]+)([^0-9a-f]*))?(([0-9a-f]+)h(.*))?(([01]+)b(.*))?(_([szhnpc]+)([^szhnpc])*)?((-?[0-9]+)([^0-9]*))?('([\S ]+)')?/i.exec(valueString);
		if(!match)
			return NaN;	// Error during parsing

		const ghex = match[3];	// 0x or $
		const ghex_empty = match[4];	// should be empty

		const ghexh = match[6];	// h
		const ghexh_empty = match[7];	// should be empty

		const gbit = match[9];	// b
		const gbit_empty = match[10];	// should be empty

		var gflags = match[12];	// _
		const gflags_empty = match[13];	// should be empty

		const gdec = match[15];	// decimal
		const gdec_empty = match[16];	// should be empty

		var gchar = match[18];	// ASCII character

		// Hex
		if(ghex) {
			if(ghex_empty)
				return NaN;
			return parseInt(ghex, 16);
		}
		if(ghexh) {
			if(ghexh_empty)
				return NaN;
			return parseInt(ghexh, 16);
		}

		// Decimal
		if(gdec) {
			if(gdec_empty)
				return NaN;
			return parseInt(gdec, 10);;
		}
		// Bits
		if(gbit) {
			if(gbit_empty)
				return NaN;
			return parseInt(gbit, 2);
		}

		// Check if status flag value
		if(gflags) {
			if(gflags_empty)
				return NaN;
			gflags = gflags.toLowerCase()
			var flags = 0;
			if(gflags.includes('s')) flags |= 0x80;
			if(gflags.includes('z')) flags |= 0x40;
			if(gflags.includes('h')) flags |= 0x10;
			if(gflags.includes('p')) flags |= 0x04;
			if(gflags.includes('n')) flags |= 0x02;
			if(gflags.includes('c')) flags |= 0x01;
			return flags;
		}

		// ASCII character
		if(gchar) {
			if(gchar.length < 1)
				return NaN;
			return gchar.charCodeAt(0);
		}

		// Unknown
		return NaN;
    }

}



Startup.main();